### PR TITLE
Add minimum possible v2 implementation

### PR DESCRIPTION
**Pull request by thewilkybarkid**

### CODE REVIEW

File: src/v2/recv.vcl [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+if (req.http.X-IIIF-Info) {
+  # Info request
+  set req.request = "HEAD";
+} else {
+  # Image request
+  set req.url = req.url "?format=" req.http.X-Fastly-IO-Format;
+}
+
+set req.http.X-Fastly-IO-URL = req.url;
```
Comment (thewilkybarkid) [inline src/v2/recv.vcl]: This shows a problem when using snippets, we want to set this at the very end of `recv` so if the URI is changed after this it's captured (eg the test suite changes the `kittehs` prefix to `cats`, but this value isn't updated). I tried to set it in `deliver`, but the IO parameters appear to have been dropped at this point.

I think we should use a single full VCL file rather than snippets so that we can include VCL before and after any custom headers, conditions etc. (Also probably easier to deploy, as Terraform will work and you only have to deploy a single file rather than many.)

File: tests/v2.test.js [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+const api = request.defaults({
+  baseUrl: 'https://api.fastly.com/service/6lQpb69X5Qt9YVt2OIMiGG/',
+  headers: {
+    Accept: 'application/json',
+    'Fastly-Key': process.env.FASTLY_API_KEY,
+  },
+});
+
+const domain = 'iotest--iiif.elifesciences.org';
```
Comment (thewilkybarkid) [inline tests/v2.test.js]: Reusing my original IO experimentation domain, should probably change to something clearer.

**Comment by thewilkybarkid** [comment]
Originally thought this library could provide separate VCL files for different IIIF versions, but I'm wondering now if we should provide a single file that can be customised to serve one _or more_ IIIF versions (relates to https://github.com/elifesciences/fastly-iiif/pull/1#discussion_r216608708, using full VCL rather than snippets).

There's not that much difference between IIIF image v2 and v3, so would end up duplicating a lot (obviously v4+ is unknown). To migrate usage from one version to another, the only option is to have the IIIF image API version in the URI (related: http://iiif.io/api/presentation/3.0/#service). So you could have a prefix of `v2` and set the IIIF version to v2, and `v3` and set the IIIF version to v3. Or always set it to a specific version (default should be v2). Means it's quite flexible and you don't have to run multiple CDNs for different IIIF versions, but makes the VCL more complex.

File: src/v2/error.vcl [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    "width": "} req.http.X-Fastly-IO-Width {",
+    "height": "} req.http.X-Fastly-IO-Height {"
+  }"};
+
+  return(deliver);
+}
+
+set obj.http.Content-Type = "text/plain; charset=us-ascii";
+
+synthetic obj.response;
```
Comment (thewilkybarkid) [inline src/v2/error.vcl]: IIIF doesn't define what to do with error messages, so following Loris's plain text approach.

File: tests/v2.test.js [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    })),
+    api.post(`version/${version}/domain`).form({
+      name: domain,
+    }),
+    api.post(`version/${version}/header`).form({
+      name: 'vary-test-run',
+      type: 'cache',
+      action: 'append',
+      dst: 'http.Vary',
+      src: '"X-Test-Run"',
```
Comment (thewilkybarkid) [inline tests/v2.test.js]: Results in 

```vcl
if (!beresp.http.Vary) {
  set beresp.http.Vary = "X-Test-Run";
} else {
  set beresp.http.Vary = beresp.http.Vary "X-Test-Run";
}
```

So should include a comma here? (`set beresp.http.Vary = ", X-Test-Run";` when it's not set would be odd, but probably doesn't cause a problem?)

Comment (thewilkybarkid) [reply tests/v2.test.js]: Might be able to do something with `regex` otherwise? (Trying to avoid writing VCL directly here.)

File: tests/v2.test.js [hunk 5]
```diff
@@ -0,0 +1,443 @@
+import expect from 'expect';
```
Comment (tamlyn) [inline tests/v2.test.js]: `expect` is automatically global within a Jest test suite so you don't need to import it.

File: tests/v2.test.js [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+      ],
+    }, requiredJson);
+
+    expect.assertions(2);
+
+    return http.get(`${prefix}/${identifier}/info.json`)
+      .then((response) => {
+        expect(response.statusCode).toBe(200);
+        expect(JSON.parse(response.body)).toEqual(json);
+      });
```
Comment (tamlyn) [inline tests/v2.test.js]: This is fine but I would favour `await`ing the request to keep the code flatter. 

Also although `expect.assertions()` (or `expect.hasAssertions()`) is essential if you are expecting in a `catch`, it's not necessary in a `then` because the only reason those expectations would not be reached is if an exception (or promise rejection) was thrown in which case the test would fail anyway.

File: tests/v2.test.js [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+    expect.assertions(1);
+
+    return http({
+      method,
+      uri: imageUri(iiifParameters),
+    })
+      .catch(exception => expect(exception).toHaveProperty('statusCode', statusCode));
+  };
+
+  const badRequest = iiifParameters => error(iiifParameters, 400);
```
Comment (tamlyn) [inline tests/v2.test.js]: You probably already know this, and it's not a problem here, but pointing it out as it's a common error: code in `describe` is executed immediately when the file is loaded. Illustration https://gist.github.com/tamlyn/dd6d54cdbe6eccd51a4bc61f5844bec4

File: tests/v2.test.js (hunk 5 above)
Comment (thewilkybarkid) [reply tests/v2.test.js]: Think that was to make my IDE happier!

File: tests/v2.test.js (hunk 7 above)
Comment (thewilkybarkid) [reply tests/v2.test.js]: Wasn't aware of that! Doesn't look a problem here, but any alternative pattern you'd recommend?

Comment (tamlyn) [reply tests/v2.test.js]: My rule of thumb is that initialising a variable or declaring a function is fine in a describe but actually calling a function should go in a `before`.

File: tests/v2.test.js (hunk 6 above)
Comment (thewilkybarkid) [reply tests/v2.test.js]: Makes sense. Tempted to use `simple: false` to avoid `catch`es too.

Comment (dryhten) [reply tests/v2.test.js]: Also, you probably already know this but if you want to change to `await` you need to make the test function `async`.

File: tests/v2.test.js [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+        content: contents,
+        main: true,
+      });
+    }));
+
+  // Send configuration requests.
+  await Promise.all(config);
+
+  // Activate the version.
+  await api.put(`version/${version}/activate`);
```
[the remark this answers is not in the document]
Comment (giorgiosironi) [reply tests/v2.test.js]: Need to poll on the `deployed` field of the `GET`?

[the remark this answers is not in the document]
Comment (giorgiosironi) [reply tests/v2.test.js]: (Or on the `active` one)

[the remark this answers is not in the document]
Comment (thewilkybarkid) [reply tests/v2.test.js]: I suspect `deployed`.

[the remark this answers is not in the document]
Comment (thewilkybarkid) [reply tests/v2.test.js]: I'm not sure what `deployed` actually means. Doesn't seem to update, and https://docs.fastly.com/api/config#version_17843bfa35e0a73c7ef0a93dfa30ee8e is confusing...

[the remark this answers is not in the document]
Comment (thewilkybarkid) [reply tests/v2.test.js]: Might have to add a response header adding the active version and poll that instead.

File: tests/v2.test.js [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+
+  // Wait for it to be deployed.
+  const checkDeployed = () => http.head('')
+    .then(response => expect(response.headers['x-fastly-config-version']).toBe(`${version}`));
```
Comment (thewilkybarkid) [inline tests/v2.test.js]: Might need to check that both the shield and the edge are up to date?

**Comment by thewilkybarkid** [comment]
Just spotted that `w,h` size parameter is required in level 0.

**Comment by thewilkybarkid** [comment]
But _not_ in 2.1. 🤔

**Comment by thewilkybarkid** [comment]
I think this is ready now, and could be a `0.1.0` release.

Interested in reviews from:

@giorgiosironi to try to use it
@diversemix for a review of the test suite
@stephenwf for a check of the IIIF compatibility
@ejthurgo for a VCL sanity check

if possible!

File: tests/v2.test.js [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+
+  test.each(paths)('%s', async (path) => {
+    const response = await http.get(path);
+
+    expect(response.statusCode).toBe(404);
+    expect(response.body).toBe('Not a IIIF path');
+  });
+});
+
+describe('Unknown versions', () => {
```
Comment (thewilkybarkid) [inline tests/v2.test.js]: Should move this to a separate file.

**Comment by giorgiosironi** [comment]
Some URLs I have tested:
- [ ] http://iotest--iiif.elifesciences.org.global.prod.fastly.net/pug-life.jpg (redirect not implemented, but it's optional)
- [x] http://iotest--iiif.elifesciences.org.global.prod.fastly.net/pug-life.jpg/info.json 
- [x] http://iotest--iiif.elifesciences.org.global.prod.fastly.net/pug-life.jpg/full/full/0/default.jpg
- [ ] http://iotest--iiif.elifesciences.org.global.prod.fastly.net/not-existing/full/full/0/default.jpg or http://iotest--iiif.elifesciences.org.global.prod.fastly.net/folder%2F1.jpg/full/full/0/default.jpg are a 403 but probably comes from the origin
- [ ] I think cropping, resizing and rotation are not supported at this moment

**Comment by giorgiosironi** [comment]
Responses to http://iotest--iiif.elifesciences.org.global.prod.fastly.net/pug-life.jpg/full/full/0/default.jpg seem to always have `Age: 0` as an header

**Comment by thewilkybarkid** [comment]
Ah, I set a `max-age` of `1` on the backend before I added the extra `Vary`. Should make it longer, and also test the header is kept.

**Comment by thewilkybarkid** [comment]
> http://iotest--iiif.elifesciences.org.global.prod.fastly.net/not-existing/full/full/0/default.jpg or http://iotest--iiif.elifesciences.org.global.prod.fastly.net/folder%2F1.jpg/full/full/0/default.jpg are a 403 but probably comes from the origin

Yep. Wonder what should happen in this case? Maybe should transform to 500 and keep the original response as a debug header? (Same with files that exist but aren't images, currently turned into a 404.)

> http://iotest--iiif.elifesciences.org.global.prod.fastly.net/pug-life.jpg (redirect not implemented, but it's optional)
> think cropping, resizing and rotation are not supported at this moment

Correct, for now it's the absolute minimum.

File: tests/environment.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+class FastlyEnvironment extends NodeEnvironment {
+  async setup() {
+    await super.setup();
+
+    const domain = process.env.FASTLY_DOMAIN;
+    const baseUrl = `https://${domain}`;
+
+    // Test client.
+    this.global.http = request.defaults({
+      baseUrl,
```
Comment (thewilkybarkid) [inline tests/environment.js]: Might want to set this explicitly to a non-shield POP (and set a `Host` header), not sure where the Travis builds are happening but might be inconsistent.

Comment (thewilkybarkid) [reply tests/environment.js]: Should also probably test shield vs edge.

File: package.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    "p-retry": "^2.0",
+    "promise-fs": "^1.3",
+    "regenerator-runtime": "^0.12.1",
+    "request": "^2.88",
+    "request-promise-native": "^1.0",
+    "uuid": "^3.3.2"
+  },
+  "scripts": {
+    "pretest": "eslint --ignore-path .gitignore .",
+    "test": "jest"
```
Comment (stephenwf) [inline package.json]: `jest --verbose` will get your tests listed, opposed to just the test-cases, will also give you millisecond timings on tests so you can see at a glance if any cases take longer

**Comment by stephenwf** [comment]
@thewilkybarkid Did some tests wrapping in a [test manifest](https://stephenwf.github.io/elife-iiif.json) for level 0 images, most viewers tend to just use the ID of the image resource in the body of the annotation, so the `info.json` [isn't loaded](http://universalviewer.io/examples/#?c=&m=&s=&cv=&manifest=https%3A%2F%2Fstephenwf.github.io%2Felife-iiif.json%20&xywh=-1%2C-1%2C1%2C0). Did a test of OpenSeadragon's vanilla handling of tile sources and came across CORS issue, so didn't get too far with that! I would definitely recommend adding in the CORS headers, even though it is technically level1 and not level0. 

Interesting use of Jest too, the set up acts almost as an installation docs.

**Comment by thewilkybarkid** [comment]
Thanks @stephenwf. I was trying to avoid adding any optional features here (for simplicity, with follow-ups to add different formats, resizes etc), but let's do CORS so we can see it working.

**Comment by thewilkybarkid** [comment]
There are some other test images in the bucket. I'll add details in the README. (Though maybe the tests should provision it?)

**Comment by stephenwf** [comment]
I can add the other test images to that manifest if you'd like. Will also put together a Presentation 2.1 manifest, so it can be tested in Mirador etc.

File: package.json (hunk 12 above)
Comment (thewilkybarkid) [reply package.json]: Only just switched to multiple files so hadn't spotted it'd changed!

**Comment by thewilkybarkid** [comment]
I think https://travis-ci.org/elifesciences/fastly-iiif/builds/429942250 has answered the shield vs edge question: no `Age` on the `info.json` must mean that it's generated at the shield, so the Travis build must be close to `dca-dc-us`.

**Comment by thewilkybarkid** [comment]
Should keep the `Age` of the original response in that case, like with the `Cache-Control`.

**Comment by giorgiosironi** [comment]
> what should happen in this case? Maybe should transform to 500 (maybe even 502?) 

Some `50x` error would be correct if we are considering this a proxy, I suspect propagating the origin's 403/404 as special cases may be more useful to the end user.

File: README.md [hunk 13]
```diff
@@ -1,2 +1,90 @@
...
+------------
+
+> You will need a Fastly service with VCL and image optimization enabled.
+
+1. Create a new configuration version.
+2. Set up a domain and host.
+3. Upload [`iiif.vcl`](iiif.vcl) as the 'main' [custom VCL](https://docs.fastly.com/vcl/custom-vcl/uploading-custom-vcl/).
+4. Upload a [configuration snippet](#configuration).
+5. Add any custom settings and snippets as required.
+6. Activate the version.
```
Comment (giorgiosironi) [inline README.md]: `7. Wait for request ... to return the header ... to confirm version has been rolled out`

File: README.md [hunk 14]
```diff
@@ -1,2 +1,90 @@
...
+
+```vcl
+sub iiif_config {
+  set req.http.X-IIIF-Version = "2";
+}
+```
+
+### Options
+
+Options are set as headers on the client request. (Note these headers and sanitised and set to their default value before `iiif_config` is called, and validated afterwards.)
```
Comment (giorgiosironi) [inline README.md]: `are sanitised`?

File: README.md [hunk 15]
```diff
@@ -1,2 +1,90 @@
...
+The version of the IIIF image API to use.
+
+Valid values:
+
+- `2`
+
+Running the tests
+-----------------
+
+> Requires Node.js8, and a Fastly service with VCL and image optimization enabled. It will create and activate a new configuration version.
```
Comment (giorgiosironi) [inline README.md]: `Node.js 8`?

File: README.md [hunk 16]
```diff
@@ -1,2 +1,90 @@
...
+
+Firstly, run `npm install`.
+
+Next, run `npm test` with the following environment variables:
+
+- `FASTLY_API_KEY` ([API token](https://docs.fastly.com/guides/account-management-and-security/using-api-tokens) with `global` scope for the Fastly service)
```
Comment (giorgiosironi) [inline README.md]: probably obvious, but from a user which is at least `Engineer`

File: README.md (hunk 13 above)
Comment (thewilkybarkid) [reply README.md]: That header's part of the test suite only.

File: README.md (hunk 16 above)
Comment (thewilkybarkid) [reply README.md]: Since it involves setting up a service to run the tests, let's just make that assumed.
